pkg/trtl/honu.go

```diff
@@ -108,8 +112,179 @@ func (h *HonuService) Delete(ctx context.Context, in *pb.DeleteRequest) (out *pb
+	}
+
+	// TODO: in order to support more complex iteration such as jumping to the next key
+	// in the page, honu needs to offer better support for leveldb iteration options.
+	// Until this is implemented, we just iterate over the prefix until we get to the
+	// start of the next page, which is extremely inefficient, especially for large
+	// datasets. [Create a story for implementing iter.Seek() in Honu]
+	var iter iterator.Iterator
+	if iter, err = h.db.Iter(prefix); err != nil {
+		log.Error().Err(err).Str("namespace", in.Namespace).Msg("could not create honu iterator")
+		return nil, status.Errorf(codes.FailedPrecondition, "could not create iterator: %s", err)
+	}
+	defer iter.Release()
+
+	for iter.Next() {
+		// Determine if we need to seek to the next page or not
+		key := iter.Key()
+		if len(cursor.NextKey) > 0 {
+			// We need to seek since there is a page token
+			// If the current key is lexicographically before the next key, then we need
+			// to continue seeking. Note that we cannot use equality here because the
+			// next key may have been deleted between requests, which means we'd seek to
+			// the end of the iteration without returning the page. Unfortunately, the
+			// lexicographic ordering that we're computing is heavily dependent on the
+			// underlying representation, so I'm just guessing with leveldb for now.
+			// TODO: this needs to be replaced with honu Seek!
+			if comparer.DefaultComparer.Compare(key, cursor.NextKey) < 0 {
+				continue
+			}
+		}
+
+		// Check if we're done iterating (e.g. at the end of the page with a next page)
+		if len(out.Values) == int(cursor.PageSize) {
+			// The current key is the next key for the next page, stop iteration
+			cursor.NextKey = key
+			break
+		}
+
+		// Otherwise append the current key value pair to the page.
+		// Fetch the metadata since it will need to be loaded for the response anyway.
+		var object *object.Object
+		if object, err = iter.Object(); err != nil {
+			log.Error().Err(err).Str("key", base64.RawURLEncoding.EncodeToString(key)).Msg("could not fetch object metadata")
+			return nil, status.Error(codes.FailedPrecondition, "database is in invalid state")
+		}
+
+		// Create the key value pair
+		pair := &pb.KVPair{}
+		if !opts.IterNoKeys {
+			pair.Key = object.Key
+			pair.Namespace = object.Namespace
+		}
+
+		if !opts.IterNoValues {
+			pair.Value = object.Data
+		}
+
+		if opts.ReturnMeta {
+			// TODO: this is duplicated code with the Get method, make it a helper function.
+			pair.Meta = &pb.Meta{
+				Key:       object.Key,
+				Namespace: object.Namespace,
+				Region:    object.Region,
+				Owner:     object.Owner,
+				Version: &pb.Version{
+					Pid:     object.Version.Pid,
+					Version: object.Version.Version,
+					Region:  object.Version.Region,
+				},
+				Parent: &pb.Version{
+					Pid:     object.Version.Parent.Pid,
+					Version: object.Version.Parent.Version,
+					Region:  object.Version.Parent.Region,
+				},
+			}
```

Comment:
This is also going to be very common code so also a good candidate for a helper function.